src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt

```diff
@@ -328,7 +328,20 @@ class ReferralService(
     referralRepository.saveAndFlush(referral)
 
     update.serviceCategoryIds?.let { serviceCategoryIds ->
-      referral.selectedServiceCategories = serviceCategoryRepository.findByIdIn(serviceCategoryIds)
+      val updatedServiceCategories = serviceCategoryRepository.findByIdIn(serviceCategoryIds)
+      // the following code looks like a strange way to do this ...
+      // however we have to be very intentional about updating the service categories,
+      // rather than replacing the collection entirely. in the former case, hibernate
+      // runs an update as you would expect. in the latter case, hibernate deletes the
+      // entity entirely and creates a new one. this causes SQL constraint violations
+      // because these service category ids are referenced in the selected complexity
```

Comment:
Not sure whether it helps but have you tried cascade delete across the relationship which violates the constraint?

Reply:
In this case we didn't want to delete those, afaik, it was semantically a no-op here but was running as a delete/insert

Reply:
yes david's correct